DOC: prep 1.9.0 release notes

tylerjereddy opened the pull request:
* transcribe entries from the [wiki page](https://github.com/scipy/scipy/wiki/Release-note-entries-for-SciPy-1.9.0) (the history of this page
will be used to update again before branching)
* draft in results from author/list scripts
* some initial manual curation of the author list via `.mailmap` adjustments
* manually update the notes based on merged PRs with `enhancement` label (for
those cases where it struck me as pertinent to mention)
* the "highlights" are still open to adjustment of course; simply
copied from the wiki for now--`meson` comes to mind depending on where
we land near the end of the week?

tylerjereddy commented on doc/release/1.9.0-notes.rst:
```diff
@@ -70,60 +145,806 @@ New features
 
+A draft developer CLI is available for SciPy, leveraging the ``doit``,
+``click`` and ``rich-click`` tools. For more details, see
+[gh-15959](https://github.com/scipy/scipy/pull/15959).
```
Maybe we can fuse the two "Other" themed sections?

rgommers replied:
Yes, good idea.

tylerjereddy commented:
CI correctly highlights a bunch of doc build warnings in the new release notes.

ev-br commented on doc/release/1.9.0-notes.rst:
```diff
@@ -26,42 +26,117 @@ For running on PyPy, PyPy3 6.0+ is required.
+- Added `scipy.stats.fit` for fitting discrete and continuous distributions
+  to data.
+- Added new spline based interpolation methods for
+  `scipy.interpolate.RegularGridInterpolator`.
```
Maybe rephrase to read "Tensor-product spline interpolation modes were added to `scipy.interpolate.RegularGridInterpolator`" or some such.

ev-br commented on doc/release/1.9.0-notes.rst:
```diff
@@ -26,42 +26,117 @@ For running on PyPy, PyPy3 6.0+ is required.
+`scipy.interpolate.RegularGridInterpolator` and its tutorial.
+- `scipy.interpolate.RegularGridInterpolator` and `scipy.interpolate.interpn`
+now accept descending ordered points.
+- ``RegularGridInterpolator`` now handles length-1 grid axes
```
Now that gh-16220 is in, let's add

- `BivariateSpline`s have grown a new method `partial_derivative` which constructs a new spline object representing a derivative of an original spline. This mirrors the corresponding functionality for univariate splines, `splder` and `BSpline.derivative`, and can substantially speed up repeated evaluation of derivatives.

ev-br reviewed (commented):
Two small tweaks --- let me know if you prefer to have them in the wiki  instead Tyler.

tupui commented on doc/release/1.9.0-notes.rst:
```diff
@@ -70,60 +145,806 @@ New features
+documentation only for a long time already.
+
+********************
+Expired Deprecations
```
@h-vetinari @j-bowhay FYI if there are other entries to add, please add them here as suggestion so that it's easier for Tyler.

h-vetinari replied:
I've tried to keep the wiki up-to-date with the issues in https://github.com/scipy/scipy/issues/15765 as they were closed. To the best of my knowledge, the list as below is complete.

tupui replied:
Thanks for confirming.

tupui commented on doc/release/1.9.0-notes.rst:
```diff
@@ -70,60 +145,806 @@ New features
+
+- ``lloyd_centroidal_voronoi_tessellation`` has been added to allow improved
+sample distributions via iterative application of Voronoi diagrams and
+centering operations
 
 Other
```
```suggestion
centering operations.

- Add `scipy.stats.qmc.PoissonDisk` to sample using the Poisson disk sampling method. It guarantees that samples are separated from each other by a given ``radius``.

Other
```

rgommers commented on doc/release/1.9.0-notes.rst:
```diff
@@ -26,42 +26,117 @@ For running on PyPy, PyPy3 6.0+ is required.
 ************
 
+Lazy access to subpackages
+==========================
```
I'd say that this doesn't belong here, it should go in the `Other changes` section, like the build system changes. `New features` is for subpackages only.

rgommers commented on doc/release/1.9.0-notes.rst:
```diff
@@ -26,42 +26,117 @@ For running on PyPy, PyPy3 6.0+ is required.
+
+Before this release, all subpackages of SciPy (`cluster`, `fft`, `ndimage`,
+etc.) had to be explicitly imported. Now, these subpackages are lazily loaded
+as soon as they are accessed, so that the following is possible:
```
add "(if desired for interactive use, it's not actually recommended for code, see `:ref:`scipy-api`).

rgommers commented on doc/release/1.9.0-notes.rst:
```diff
@@ -26,42 +26,117 @@ For running on PyPy, PyPy3 6.0+ is required.
+- Added new spline based interpolation methods for
+  `scipy.interpolate.RegularGridInterpolator`.
+- A new global optimizer (DIviding RECTangles algorithm)
+  `scipy.optimize.direct`
```
I have the Meson release note changes in the PR to change to it by default - but not a highlight. Would be good to add that here indeed.

rgommers commented on doc/release/1.9.0-notes.rst:
```diff
@@ -26,42 +26,117 @@ For running on PyPy, PyPy3 6.0+ is required.
 ============================
 
+`scipy.fft` improvements
+============================
```
minor: underline length needs to match title length

rgommers commented on doc/release/1.9.0-notes.rst:
```diff
@@ -26,42 +26,117 @@ For running on PyPy, PyPy3 6.0+ is required.
 ===========================
-
-
+- Add a `copy` parameter to `scipy.sparce.csgraph.laplacian`. Using inplace
```
formatting: double backticks for `copy`

rgommers replied:
and same for `copy=False` below

rgommers commented on doc/release/1.9.0-notes.rst:
```diff
@@ -70,60 +145,806 @@ New features
+pdf and cdf calculation. More information can be found in the documentation
+here (to be linked).
+
+- Added `scipy.stats.fit` for fitting discrete and continuous distributions to
```
Just noticing this now, `fit` is a _very_ generic name. I can't remember it being proposed/discussed. Will comment on gh-15436.

ev-br commented on doc/release/1.9.0-notes.rst:
```diff
@@ -70,60 +145,806 @@ New features
+`scipy.integrate.quad`, have been marked as `recursive`. This should fix rare
+issues in multivariate integration (`nquad` and friends) and obviate the need
+for compiler-specific compile flags (`/recursive` for ifort etc). Please file
+an issue if this change turns out problematic for you.
```
Also FITPACK routines in `scipy.interpolate`, which power `splrep`, `splev` etc, and `*UnivariateSpline` and `*BivariateSpline` classes.

mdhaber commented on doc/release/1.9.0-notes.rst:
```diff
@@ -70,60 +145,806 @@ New features
+- Remove inheritance to `QMCEngine` in `MultinomialQMC` and
+`MultivariateNormalQMC`. It removes the methods `fast_forward` and `reset`.
+- Init of `MultinomialQMC` now require the number of trials with `n_trials`.
+Hence, `MultinomialQMC.random` output has now the correct shape ``(n, pvals)``.
```
Email [here](https://mail.python.org/archives/list/scipy-dev@python.org/thread/QEMS6VZSBHTRB6OVSDBULJNNOE7PFAOV/#QEMS6VZSBHTRB6OVSDBULJNNOE7PFAOV), discussion [here](https://github.com/scipy/scipy/pull/15945). This was definitely a bug.

mdhaber commented on doc/release/1.9.0-notes.rst:
```diff
@@ -70,60 +145,806 @@ New features
 Pull requests for 1.9.0
 ***********************
 
+* `#354 <https://github.com/scipy/scipy/pull/354>`__: Faster expm
```
```suggestion
```
This wasn't actually merged; it was superseded. Perhaps the code that generates this list could use a little tweak? Not too important, though.

mdhaber commented on doc/release/1.9.0-notes.rst:
```diff
@@ -70,60 +145,806 @@ New features
+- added ``qrvs`` method to ``NumericalInversePolynomial`` to match the
+situation for ``NumericalInverseHermite``
+
+- faster random variate generation for ``gennorm``
```
```suggestion
- faster random variate generation for ``gennorm`` and ``nakagami``
```
gh-15360

mdhaber commented on doc/release/1.9.0-notes.rst:
```diff
@@ -70,60 +145,806 @@ New features
+ndarray with the ``axis`` dimension removed.
+
+- Replaced implementation of `scipy.stats.ncf` with the implementation from
+Boost.
```
```suggestion
Boost for improved reliability.
```

tylerjereddy commented:
I think I addressed all the comments in the PR so far, and `make html-scipyorg` passes locally (using the old `setuptools` approach). I've also included a fix to `gh_lists.py` that ignores PRs that were never actually merged (i.e., the closed PRs @mdhaber pointed out are no longer appended now). You can see the revisions in the second commit in the PR, `307443`. I also rebased to pull in the latest `meson` stuff just merged.

There will still be one or two more merges to include here and I'll check the wiki again before finalizing here/branching. I'll check the html version of the relnotes produced locally as well if I can.

tylerjereddy commented:
From the html preview locally, I'd say one or two sections below the fold look incorrectly formatted (`Expired Deprecations`, `Lazy access to subpackages`; the former looks wrong, the latter my impression is that there may have been intention for the snippets not to be inline like that?), but the rest is reasonably professional looking. Perhaps less severe, but the `levy_stable` bullet point is so verbose that it almost looks incorrectly formatted, but maybe that one is "ok".

<details>

![image](https://user-images.githubusercontent.com/7903078/170844772-b1c9e840-3f31-452b-a7b5-8008522c798a.png)

![image](https://user-images.githubusercontent.com/7903078/170844790-0a42f663-c251-4f22-a10e-b6069f7abc3b.png)



</details>

WarrenWeckesser commented on doc/release/1.9.0-notes.rst:
```diff
@@ -70,46 +139,213 @@ New features
+********************
+There is an ongoing effort to follow through on long-standing deprecations.
+The following previously deprecated features are affected:
+- Object arrays in sparse matrices now raise an error
```
You need a blank line before the beginning of the bulleted list.  (I was going to link to the comment about this in the "missing bits" document, but apparently the link to that document in the rendered docs was broken by some recent reshuffling of the docs files. I see if I can fix that and submit a PR.)

WarrenWeckesser replied:
Here's the link to the relevant section of the docs from a PR that fixes the reference to the "missing bits" file: [Bulleted lists](https://output.circle-artifacts.com/output/job/7265df6e-f1f4-4e37-9ca6-2cd6a7b903f1/artifacts/0/html-scipyorg/dev/contributor/missing-bits.html#bulleted-lists)

tupui commented:
@tylerjereddy before branching, there is an issue with the version switcher url (saying this here in case you wanted to branch before I make a PR tmr). Since the URL of the doc changed/got fixed, it also need to be adapted on the switcher. And we need to add 1.8.1.

mdhaber commented on doc/release/1.9.0-notes.rst:
```diff
@@ -70,46 +139,213 @@ New features
+- Remove inheritance to `QMCEngine` in `MultinomialQMC` and
+  `MultivariateNormalQMC`. It removes the methods `fast_forward` and `reset`.
+- Init of `MultinomialQMC` now require the number of trials with `n_trials`.
+  Hence, `MultinomialQMC.random` output has now the correct shape ``(n, pvals)``.
```
```suggestion
  Hence, `MultinomialQMC.random` output has now the correct shape ``(n, pvals)``.
- Several function-specific warnings (``F_onewayConstantInputWarning``,
  ``F_onewayBadInputSizesWarning``, ``PearsonRConstantInputWarning``, 
  ``PearsonRNearConstantInputWarning``, ``SpearmanRConstantInputWarning``, and
  ``BootstrapDegenerateDistributionWarning``) have been replaced with more
  general warnings.
```

tylerjereddy commented:
This is the last open milestoned PR, so branching is imminent. I'm iterating on some final relnotes updates/fixes locally now.

mdhaber commented on doc/release/1.9.0-notes.rst:
```diff
@@ -26,90 +26,329 @@ For running on PyPy, PyPy3 6.0+ is required.
+- Add a `bits` parameter to `scipy.stats.qmc.Sobol`. It allows to use from 0
+  to 64 bits to compute the sequence. Default is ``None`` which corresponds to
+  30 for backward compatibility. Using a higher value allow to sample more
+  points. Note: it does not correspond to the output type.
```
```suggestion
  points. Note: ``bits`` does not affect the output dtype.
```

tylerjereddy commented:
@mdhaber I committed your most recent suggestion, feel free to self-commit those small ones though, I hope to merge this and branch if I see that the CI is passing when I check a bit later.

mdhaber commented on doc/release/1.9.0-notes.rst:
```diff
@@ -26,90 +26,329 @@ For running on PyPy, PyPy3 6.0+ is required.
+  ``alpha`` is deprecated. Keyword ``alpha`` is replaced with keyword
+  ``confidence``.
+- The ``'simplex'``, ``'revised simplex'``, and ``'interior-point'`` methods
+  of `scipy.optimize.linprog` are deprecated. Methods ``highs``, ``highs-ds``,
```
```suggestion
  of ``scipy.optimize.linprog`` are deprecated. Methods ``highs``, ``highs-ds``,
```
I'm not sure if this will link properly.

tylerjereddy commented:
Updated the lists to account for last-minute gh-16241

mdhaber commented on doc/release/1.9.0-notes.rst:
```diff
@@ -26,90 +26,329 @@ For running on PyPy, PyPy3 6.0+ is required.
-
-`scipy.special` improvements
-============================
+- Add a `order` parameter to `scipy.spatial.transform.Rotation.from_quat` and
+  `scipy.spatial.transform.Rotation.as_quat` to specify quaternion format.
```
```suggestion
- Add an ``order` `parameter to `scipy.spatial.transform.Rotation.from_quat` 
  and `scipy.spatial.transform.Rotation.as_quat` to specify quaternion format.
```

mdhaber commented on doc/release/1.9.0-notes.rst:
```diff
@@ -26,90 +26,329 @@ For running on PyPy, PyPy3 6.0+ is required.
+  size limitations, it makes it possible to perform similar tests using arbitrary
+  statistics and distributions.
+
+- Several ``scipy.stats`` functions support new ``axis`` (integer or tuple of
```
```suggestion
- Several `scipy.stats` functions support new ``axis`` (integer or tuple of
```

mdhaber commented on doc/release/1.9.0-notes.rst:
```diff
@@ -26,90 +26,329 @@ For running on PyPy, PyPy3 6.0+ is required.
+  integers) and ``nan_policy`` ('raise', 'omit', or 'propagate'), and
+  ``keepdims`` arguments.
+  These functions also support masked arrays as inputs, even if they do not have
+  a ``scipy.stats.mstats`` counterpart. Edge cases for multidimensional arrays,
```
```suggestion
  a `scipy.stats.mstats` counterpart. Edge cases for multidimensional arrays,
```

mdhaber commented on doc/release/1.9.0-notes.rst:
```diff
@@ -26,90 +26,329 @@ For running on PyPy, PyPy3 6.0+ is required.
+  or ``highs-ipm`` should be used in new code.
+- Support for non-numeric arrays has been deprecated from ``stats.mode``.
+  ``pandas.DataFrame.mode`` can be used instead.
+- The function ``spatial.distance.kulsinski`` has been deprecated in favor
+  of ``spatial.distance.kulczynski1``.
```
```suggestion
- The function `spatial.distance.kulsinski` has been deprecated in favor
  of `spatial.distance.kulczynski1`.
```

mdhaber commented on doc/release/1.9.0-notes.rst:
```diff
@@ -26,90 +26,329 @@ For running on PyPy, PyPy3 6.0+ is required.
+
+- Object arrays in sparse matrices now raise an error
+- Inexact indices into sparse matrices now raise an error
+- Passing ``radius=None`` to ``spatial.SphericalVoronoi`` now raises an error
+  (not adding ``radius`` defaults to 1, as before)
```
```suggestion
- Passing ``radius=None`` to `scipy.spatial.SphericalVoronoi` now raises an
  error (not adding ``radius`` defaults to 1, as before)
```

mdhaber commented on doc/release/1.9.0-notes.rst:
```diff
@@ -26,90 +26,329 @@ For running on PyPy, PyPy3 6.0+ is required.
+- The ``_rvs`` method of statistical distributions now requires a ``size``
+  parameter
+- Passing a ``fillvalue`` that cannot be cast to the output type in
+  ``signal.convolve2d`` now raises an error
```
```suggestion
  `scipy.signal.convolve2d` now raises an error
```

mdhaber commented on doc/release/1.9.0-notes.rst:
```diff
@@ -26,90 +26,329 @@ For running on PyPy, PyPy3 6.0+ is required.
+  Similarly, while masked elements of masked arrays are still ignored, the
+  output will be a scalar or ``np.ndarray`` rather than a masked array with
+  ``mask=False``.
+- The default method of ``scipy.optimize.linprog`` is now ``'highs'``, not
```
```suggestion
- The default method of `scipy.optimize.linprog` is now ``'highs'``, not
```

mdhaber commented on doc/release/1.9.0-notes.rst:
```diff
@@ -26,90 +26,329 @@ For running on PyPy, PyPy3 6.0+ is required.
+  parameter
+- Passing a ``fillvalue`` that cannot be cast to the output type in
+  ``signal.convolve2d`` now raises an error
+- The ``spatial.distance`` now enforce that the input vectors are
```
```suggestion
- `scipy.spatial.distance` now enforces that the input vectors are
```

mdhaber commented on doc/release/1.9.0-notes.rst:
```diff
@@ -26,90 +26,329 @@ For running on PyPy, PyPy3 6.0+ is required.
+  * ``array`` is a numpy array (backward compatible default);
+  * ``function`` is a pointer to a lambda-function evaluating the
+  Laplacian-vector or Laplacian-matrix product;
+  * ``lo`` results in the format of the `LinearOperator`.
```
```suggestion
  * ``lo`` results in the format of the ``LinearOperator``.
```

mdhaber reviewed (approved):
I'm sure there are many single VS double backtick issues still present, but I should probably stop so as not to hold up release.

tylerjereddy commented:
Ok, the docs are building/passing in CI so in it goes. I don't think we need to flush Azure an Nth time for release notes typo fixes, etc.

rgommers commented on doc/release/1.9.0-notes.rst:
```diff
@@ -152,14 +391,674 @@ For more information on the introduction of Meson support in SciPy, see
 
 ***********************
 Issues closed for 1.9.0
 ***********************
```
Indeed, see `tools/gh_lists.py`.

Grabbing issues based on date of closure isn't going to work, because a lot are closed because fixed in older releases, invalid reports, etc.

As of now, you should set milestones on issues that you close. A way to improve the script could be to parse commit messages for "Closes xxx", but even that isn't going to be foolproof since I often see that people fail to add that info to commit messages and only add it to PR descriptions.

ev-br replied:
A possible automation could be to auto-add a milestone to an issue which is closed by a PR  with a milestone. Maybe that's a feature request to github or maybe it's possible to do with their api.

mdhaber replied:
I started to do this for issues I was involved with. I thought I should make sure I'm doing the right thing before going further. For instance, I added the milestone to gh-14622 because it was a legitimate issue that was closed by a PR, whereas I did not add a milestone to gh-15160 because it was reported due to a misunderstanding. LMK if I should keep going this way.

rgommers replied:
Yes, that sounds right, thanks Matt!

mdhaber replied:
OK, done. I added a lot of milestones. 
I accidentally created a 1.90 milestone in the process. Do you know how to remove that from the list?

rgommers replied:
Thanks Matt. I deleted the 1.90 milestone.
